packages/menu-surface/index.js

```diff
@@ -0,0 +1,294 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+import {MDCMenuSurfaceFoundation, Corner} from '@material/menu-surface';
+
+class MenuSurface extends React.Component {
+
+  menuSurfaceElement_ = React.createRef();
+  previousFocus_ = null;
+  foundation_ = null;
+
+  state = {
+    transformOrigin: '',
+    maxHeight: '',
+    styleLeft: null,
+    styleRight: null,
+    styleTop: null,
+    styleBottom: null,
+    classList: new Set(),
+  };
+
+  componentDidMount() {
+    const {anchorCorner, anchorMargin, coordinates, fixed, open, quickOpen} = this.props;
+    this.handleWindowClick_ = (evt) => this.foundation_.handleBodyClick(evt);
+
+    this.registerWindowClickListener_ = () => window.addEventListener('click', this.handleWindowClick_);
+    this.deregisterWindowClickListener_ = () => window.removeEventListener('click', this.handleWindowClick_);
+
+    this.foundation_ = new MDCMenuSurfaceFoundation(this.adapter);
+    this.foundation_.init();
+
+    this.hoistToBody();
+    this.foundation_.setFixedPosition(fixed);
+    if (coordinates) {
+      this.setCoordinates();
+    }
+    if (anchorCorner) {
+      this.foundation_.setAnchorCorner(anchorCorner);
+    }
+    if (anchorMargin) {
+      this.foundation_.setAnchorMargin(anchorMargin);
+    }
+    if (quickOpen) {
+      this.foundation_.setQuickOpen(quickOpen);
+    }
+
+    if (open) {
+      this.open_();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.open !== prevProps.open) {
+      this.open_();
+    }
+    if (this.props.coordinates !== prevProps.coordinates) {
+      this.setCoordinates();
+    }
+    if (this.props.anchorCorner !== prevProps.anchorCorner) {
+      this.foundation_.setAnchorCorner(this.props.anchorCorner);
+    }
+    if (this.props.anchorMargin !== prevProps.anchorMargin) {
+      this.foundation_.setAnchorMargin(this.props.anchorMargin);
+    }
+    if (this.props.quickOpen !== prevProps.quickOpen) {
+      this.foundation_.setQuickOpen(this.props.quickOpen);
+    }
+  }
+
+  componentWillUnmount() {
+    this.deregisterWindowClickListener_();
+    this.foundation_.destroy();
+  }
+
+  hoistToBody() {
+    // this deviatees from the mdc web version.
+    // here we force the menu to hoist, and require either
+    // this.props.(x,y) or this.props.anchorElement.
+    if (!(this.menuSurfaceElement_ && this.menuSurfaceElement_.current)) return;
```

Comment:
I think we can remove the null check for `this.menuSurfaceElement_`

Reply:
I think its safer to keep it in. `adapter.restoreFocus` is called within `foundation.close()`. `foundation.close` is called in `handleBodyClick` and `handleKeydown`, which are async events. There is the possibility that the MenuSurface is unmounted before a keydown event or a click event occurs, which would cause an error. 

What do you think?